gateway/src/p2p.rs

```diff
@@ -0,0 +1,237 @@
+            meta_exchange: meta_exchange::create_behaviour(),
+        };
+
+        let swarm = Swarm::new(
+            DnsTransport::system(transport).unwrap().boxed(),
+            behaviour,
+            peer_id,
+            SwarmConfig::with_tokio_executor()
+                .with_idle_connection_timeout(Duration::from_secs(60)),
+        );
+
+        let (tx_request, rx_request) = mpsc::unbounded_channel::<Request>();
+        let (tx_response, rx_response) = mpsc::unbounded_channel::<Response>();
+
+        Ok(P2P {
+            node_urls,
+            peer_id,
+            swarm,
+            channels: Channels {
+                tx_request,
+                rx_request,
+                tx_response,
+                rx_response: Value::Here(rx_response),
+            },
+        })
+    }
+
+    /// starts listening and performs all the bindings but doesn't react yeat
+    pub fn prepare(&mut self) -> Result<(), Box<dyn std::error::Error>> {
+        for url in &self.node_urls {
+            let addr: Multiaddr = url
+                .parse()
+                .map_err(|e| format!("parse url: {}: {}", url, e))?;
+            debug!("Dialing {addr} …");
+            self.swarm.dial(addr)?;
+        }
+
+        Ok(())
+    }
+
+    // Gets p2p channels that will be used for communication
+    // can be called only once
+    pub fn interface_channels(&mut self) -> P2PChannels {
+        let mut moved = Value::Moved;
+        std::mem::swap(&mut moved, &mut self.channels.rx_response);
+
+        match moved {
+            Value::Moved => {
+                todo!(
+                    "already moved. This fn can be called only once. Maybe return proper error here?"
+                )
+            }
+            Value::Here(rx_resp) => {
+                return P2PChannels {
+                    tx_request: self.channels.tx_request.clone(),
+                    rx_response: rx_resp,
+                };
+            }
+        }
+    }
+
+    /// blocking operation, so you might want to spawn it on a separate thread
+    /// after calling this - channels at `interface_channels` will start to send messages
+    /// TODO: add stop/finish channel
+    pub async fn start_event_loop(self) {
+        let mut maroon_peer_ids = HashSet::<PeerId>::new();
+        let mut swarm = self.swarm;
+
+        let mut rx_request = self.channels.rx_request;
+        let tx_response = self.channels.tx_response;
+        loop {
+            tokio::select! {
+                Some(request) = rx_request.recv() =>  {
+                    for peer_id in &maroon_peer_ids {
+                        debug!("Sending request to {}", peer_id);
+                        let _request_id = swarm.behaviour_mut().request_response.send_request(peer_id, request.clone());
+                    }
+                },
+                event = swarm.select_next_some()=>{
+                    match event{
```

Comment:
Nit: missing spaces here too. (Perhaps implement these cases as lambdas before the outer `select!` ?